map_claims.go

```diff
@@ -2,150 +2,107 @@ package jwt
 
 import (
 	"encoding/json"
-	"errors"
-	"time"
-	// "fmt"
 )
 
 // MapClaims is a claims type that uses the map[string]interface{} for JSON decoding.
 // This is the default claims type if you don't supply one
 type MapClaims map[string]interface{}
 
-// VerifyAudience Compares the aud claim against cmp.
-// If required is false, this method will return true if the value matches or is unset
-func (m MapClaims) VerifyAudience(cmp string, req bool) bool {
-	var aud []string
-	switch v := m["aud"].(type) {
-	case string:
-		aud = append(aud, v)
-	case []string:
-		aud = v
-	case []interface{}:
-		for _, a := range v {
-			vs, ok := a.(string)
-			if !ok {
-				return false
-			}
-			aud = append(aud, vs)
-		}
-	}
-	return verifyAud(aud, cmp, req)
+// GetExpirationTime implements the Claims interface.
+func (m MapClaims) GetExpirationTime() (*NumericDate, error) {
+	return m.ParseNumericDate("exp")
 }
 
-// VerifyExpiresAt compares the exp claim against cmp (cmp <= exp).
-// If req is false, it will return true, if exp is unset.
-func (m MapClaims) VerifyExpiresAt(cmp int64, req bool) bool {
-	cmpTime := time.Unix(cmp, 0)
-
-	v, ok := m["exp"]
-	if !ok {
-		return !req
-	}
-
-	switch exp := v.(type) {
-	case float64:
-		if exp == 0 {
-			return verifyExp(nil, cmpTime, req)
-		}
+// GetNotBefore implements the Claims interface.
+func (m MapClaims) GetNotBefore() (*NumericDate, error) {
+	return m.ParseNumericDate("nbf")
+}
 
-		return verifyExp(&newNumericDateFromSeconds(exp).Time, cmpTime, req)
-	case json.Number:
-		v, _ := exp.Float64()
+// GetIssuedAt implements the Claims interface.
+func (m MapClaims) GetIssuedAt() (*NumericDate, error) {
+	return m.ParseNumericDate("iat")
+}
 
-		return verifyExp(&newNumericDateFromSeconds(v).Time, cmpTime, req)
-	}
+// GetAudience implements the Claims interface.
+func (m MapClaims) GetAudience() (ClaimStrings, error) {
+	return m.ParseClaimsString("aud")
+}
 
-	return false
+// GetIssuer implements the Claims interface.
+func (m MapClaims) GetIssuer() (string, error) {
+	return m.ParseString("iss")
 }
 
-// VerifyIssuedAt compares the exp claim against cmp (cmp >= iat).
-// If req is false, it will return true, if iat is unset.
-func (m MapClaims) VerifyIssuedAt(cmp int64, req bool) bool {
-	cmpTime := time.Unix(cmp, 0)
+// GetSubject implements the Claims interface.
+func (m MapClaims) GetSubject() (string, error) {
+	return m.ParseString("sub")
+}
 
-	v, ok := m["iat"]
+// ParseNumericDate tries to parse a key in the map claims type as a number
+// date. This will succeed, if the underlying type is either a [float64] or a
+// [json.Number]. Otherwise, nil will be returned.
+func (m MapClaims) ParseNumericDate(key string) (*NumericDate, error) {
```

Comment:
```suggestion
func (m MapClaims) parseNumericDate(key string) (*NumericDate, error) {
```

Can this (and the other helpers) be unexported, feels like an internal implementation detail .. we can always export them if needed.

Reply:
Reading the helpers a bit more, I think we definitely don't want to expose these because there is unexpected behaviour. E.g.,

```
v, ok := m[key]
if !ok {
	return nil, nil
}
```

Should this return an `ErrClaimNotFound` error?

Reply:
That depends on where you want to catch the "non-existence". They way it is built is that this function returns a `nil` numeric date if the date field is not set. This is not really an error, because the date fields such as `exp` and such are optional. However, if the field is populated with a non-date type, then an error must be thrown, otherwise this could lead to problems in the token verification step.

Reply:
I can unexpert them for now